File: backend/queryskiff/workspace.py
"""Multi-dataset query workspace (HEL-112).

A workspace is a per-request list of (dataset, alias) entries. The browser
sends opaque dataset ids + aliases; the server resolves each id (re-validated
against the allow-list — a forged id fails exactly like the single-dataset
path), validates every alias, and exposes each dataset to SQL ONLY under its
alias. Storage paths and credentials never appear in the SQL or the browser.

Aliases become CTE/view names, so they must be plain lowercase identifiers and
must not collide with each other or shadow SQL keywords in confusing ways.
The legacy single-dataset flow is the degenerate one-entry workspace with the
implicit alias `data`.
"""
from __future__ import annotations

import re
from dataclasses import dataclass

from . import datasets
from .datasets import Dataset

_ALIAS_RE = re.compile(r"^[a-z][a-z0-9_]{0,29}$")
# names that would confuse validation or shadow engine namespaces
_RESERVED = {
    "select", "from", "where", "join", "on", "with", "as", "group", "order",
    "by", "limit", "union", "all", "distinct", "having", "case", "when",
    "then", "else", "end", "and", "or", "not", "null", "true", "false",
    "table", "values", "minio", "pg", "oracle", "system",
}
MAX_DATASETS = 8


class WorkspaceError(ValueError):
    pass


@dataclass(frozen=True)
class Entry:
    dataset: Dataset
    alias: str


def resolve_entries(payload: list[dict]) -> list[Entry]:
    """[{dataset_id, alias}] -> validated entries. Raises WorkspaceError with a
    browser-safe message on any forged id, bad alias, or collision."""
    if not payload:
        raise WorkspaceError("workspace needs at least one dataset")
    if len(payload) > MAX_DATASETS:
        raise WorkspaceError(f"at most {MAX_DATASETS} datasets per workspace")
    entries: list[Entry] = []
    seen_aliases: set[str] = set()
    seen_ids: set[str] = set()
    for item in payload:
        alias = str(item.get("alias", "")).strip().lower()
        if not _ALIAS_RE.match(alias):
            raise WorkspaceError(
                f"invalid alias {alias!r} (lowercase letters/digits/underscore, "
                f"must start with a letter, max 30 chars)")
        if alias in _RESERVED:
            raise WorkspaceError(f"alias {alias!r} is reserved")
        if alias in seen_aliases:
            raise WorkspaceError(f"duplicate alias {alias!r}")
        try:
            ds = datasets.resolve_id(str(item.get("dataset_id", "")))
        except ValueError as exc:
            raise WorkspaceError(f"dataset for alias {alias!r}: {exc}") from exc
        if ds.dataset_id in seen_ids:
            raise WorkspaceError(f"dataset appears twice (alias {alias!r})")
        seen_aliases.add(alias)
        seen_ids.add(ds.dataset_id)
        entries.append(Entry(dataset=ds, alias=alias))
    return entries


def join_hints(schemas: dict[str, list[dict]]) -> list[dict]:
    """Column-compatibility hints across workspace datasets: same column name in
    two aliases, with a type-compatibility note. Hints only — never enforced."""
    _NUM = {"tinyint", "smallint", "integer", "bigint", "float", "double",
            "real", "decimal", "hugeint", "ubigint", "uinteger"}

    def kind(t: str) -> str:
        t = t.lower().split("(")[0].strip()
        if t in _NUM:
            return "numeric"
        if "timestamp" in t or t == "date":
            return "temporal"
        if "char" in t or t in ("varchar", "text", "string"):
            return "text"
        return t

    cols: dict[str, list[tuple[str, str]]] = {}
    for alias, schema in schemas.items():
        for col in schema:
            name = str(col.get("column_name") or col.get("name") or "").lower()
            ctype = str(col.get("column_type") or col.get("type") or "")
            if name:
                cols.setdefault(name, []).append((alias, ctype))
    hints = []
    for name, owners in sorted(cols.items()):
        if len(owners) < 2:
            continue
        kinds = {kind(t) for _, t in owners}
        hints.append({
            "column": name,
            "aliases": [{"alias": a, "type": t} for a, t in owners],
            "compatible": len(kinds) == 1,
            "note": None if len(kinds) == 1 else
                    f"type kinds differ ({', '.join(sorted(kinds))}) — cast before joining",
        })
    return hints


def starter_join_sql(entries: list[Entry], hints: list[dict]) -> str:
    """Editable starter SQL: joins the first two aliases on the best hint
    (compatible common column), else a cross-join skeleton with a TODO."""
    if len(entries) < 2:
        return f"SELECT * FROM {entries[0].alias} LIMIT 100"
    a, b = entries[0].alias, entries[1].alias
    best = next((h for h in hints if h["compatible"]
                 and {x["alias"] for x in h["aliases"]} >= {a, b}), None)
    if best:
        c = best["column"]
        return (f"SELECT *\nFROM {a}\nJOIN {b} ON {a}.{c} = {b}.{c}\nLIMIT 100")
    return (f"SELECT *\nFROM {a}\nJOIN {b} ON /* TODO: join condition */ 1=1\nLIMIT 100")

File: backend-jvm/src/main/kotlin/com/queryskiff/workspace/Workspace.kt
package com.queryskiff.workspace

import com.queryskiff.datasets.Datasets

/**
 * HEL-95 phase 4: JVM port of `queryskiff.workspace` (HEL-112) — the
 * multi-dataset query workspace. Per-request (dataset, alias) entries; the
 * server resolves opaque ids (forged ids fail exactly like the single-dataset
 * path), validates every alias, and exposes each dataset to SQL only under its
 * alias. Behavior mirrors the Python module 1:1 (messages included) so the
 * HTTP workspace contract passes unchanged.
 */
object Workspace {
    private val ALIAS_RE = Regex("^[a-z][a-z0-9_]{0,29}$")
    private val RESERVED = setOf(
        "select", "from", "where", "join", "on", "with", "as", "group", "order",
        "by", "limit", "union", "all", "distinct", "having", "case", "when",
        "then", "else", "end", "and", "or", "not", "null", "true", "false",
        "table", "values", "minio", "pg", "oracle", "system",
    )
    const val MAX_DATASETS = 8

    class WorkspaceError(message: String) : Exception(message)

    data class Entry(val dataset: Datasets.Dataset, val alias: String)

    /** [{dataset_id, alias}] -> validated entries; browser-safe errors. */
    fun resolveEntries(
        payload: List<Map<String, Any?>>,
        allowedBuckets: Collection<String>,
    ): List<Entry> {
        if (payload.isEmpty()) throw WorkspaceError("workspace needs at least one dataset")
        if (payload.size > MAX_DATASETS)
            throw WorkspaceError("at most $MAX_DATASETS datasets per workspace")
        val entries = mutableListOf<Entry>()
        val seenAliases = mutableSetOf<String>()
        val seenIds = mutableSetOf<String>()
        for (item in payload) {
            val alias = (item["alias"]?.toString() ?: "").trim().lowercase()
            if (!ALIAS_RE.matches(alias))
                throw WorkspaceError(
                    "invalid alias '$alias' (lowercase letters/digits/underscore, " +
                    "must start with a letter, max 30 chars)")
            if (alias in RESERVED) throw WorkspaceError("alias '$alias' is reserved")
            if (alias in seenAliases) throw WorkspaceError("duplicate alias '$alias'")
            val ds = try {
                Datasets.resolveId(item["dataset_id"]?.toString() ?: "", allowedBuckets)
            } catch (e: Datasets.DatasetError) {
                throw WorkspaceError("dataset for alias '$alias': ${e.message}")
            }
            if (ds.datasetId in seenIds)
                throw WorkspaceError("dataset appears twice (alias '$alias')")
            seenAliases += alias
            seenIds += ds.datasetId
            entries += Entry(ds, alias)
        }
        return entries
    }

    private val NUMERIC = setOf(
        "tinyint", "smallint", "integer", "bigint", "float", "double",
        "real", "decimal", "hugeint", "ubigint", "uinteger",
    )

    private fun kind(t: String): String {
        val k = t.lowercase().substringBefore("(").trim()
        return when {
            k in NUMERIC -> "numeric"
            "timestamp" in k || k == "date" -> "temporal"
            "char" in k || k in setOf("varchar", "text", "string") -> "text"
            else -> k
        }
    }

    data class HintOwner(val alias: String, val type: String)
    data class Hint(val column: String, val aliases: List<HintOwner>,
                    val compatible: Boolean, val note: String?)

    /** Column-compatibility hints across workspace schemas (hints only). */
    fun joinHints(schemas: Map<String, List<Map<String, Any?>>>): List<Hint> {
        val cols = linkedMapOf<String, MutableList<HintOwner>>()
        for ((alias, schema) in schemas) {
            for (col in schema) {
                val name = (col["column_name"] ?: col["name"] ?: "").toString().lowercase()
                val ctype = (col["column_type"] ?: col["type"] ?: "").toString()
                if (name.isNotEmpty())
                    cols.getOrPut(name) { mutableListOf() }.add(HintOwner(alias, ctype))
            }
        }
        return cols.entries.sortedBy { it.key }.mapNotNull { (name, owners) ->
            if (owners.size < 2) return@mapNotNull null
            val kinds = owners.map { kind(it.type) }.toSet()
            Hint(
                column = name, aliases = owners, compatible = kinds.size == 1,
                note = if (kinds.size == 1) null
                       else "type kinds differ (${kinds.sorted().joinToString(", ")}) — cast before joining",
            )
        }
    }

    /** Editable starter SQL: joins the first two aliases on the best hint. */
    fun starterJoinSql(entries: List<Entry>, hints: List<Hint>): String {
        if (entries.size < 2) return "SELECT * FROM ${entries[0].alias} LIMIT 100"
        val a = entries[0].alias
        val b = entries[1].alias
        val best = hints.firstOrNull { h ->
            h.compatible && h.aliases.map { it.alias }.toSet().containsAll(setOf(a, b))
        }
        return if (best != null) {
            val c = best.column
            "SELECT *\nFROM $a\nJOIN $b ON $a.$c = $b.$c\nLIMIT 100"
        } else {
            "SELECT *\nFROM $a\nJOIN $b ON /* TODO: join condition */ 1=1\nLIMIT 100"
        }
    }
}
